Add Kernel Misc Probe

vincentmli opened the pull request:
Add Kernel Misc Probe
    
Add kernel misc feature probe which include kernel BPF large instruction
support information. It allows Cilium features to probe kernel large instruction
    
Signed-off-by: Vincent Li <vincent.mc.li@gmail.com>

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [ ] All code is covered by unit and/or runtime tests where feasible.
- [ ] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

<!-- Description of change -->



```release-note
Add Kernel Misc Probe
```

joestringer commented on pkg/datapath/linux/requirements.go:
```diff
@@ -135,6 +135,12 @@ func CheckMinRequirements() {
+		// Warn about kernel not supporting large 1M instruction
+		if err := probeManager.ProbeKernelLargeInstruction(); err != nil {
+			log.WithError(err).Warn("BPF check: NOT OK")
+		}
```
I think we need to consider exactly what we're trying to communicate here.

As-is, this code will *always* warn any user running roughly version ~5.2 of Linux or earlier. There is no impact to those users, it's just noise. Many distributions rely on Linux 4.14, 4.15, 4.19. If users see this warning, those users on those kernels may be under the impression that something is broken, but nothing is broken. Also, what is the user supposed to do about this? `BPF check: NOT OK` is pretty cryptic as a message.

I'm not sure, do we need this particular hunk of code at all?

vincentmli replied:
@joestringer This is to address https://github.com/cilium/cilium/pull/17370 exceeding 4.19 kernel  4K instruction limit, I thought you mentioned to send a separate PR for kernel limit probing and warning https://github.com/cilium/cilium/pull/17370#issuecomment-932579022, or I misunderstand you :)

joestringer replied:
The PR makes sense to integrate the code (so that others like @pchaigno can comment & also build on top), but warning every user on an older kernel doesn't make sense.

vincentmli replied:
I see what you mean, I can  warn only when https://github.com/cilium/cilium/pull/17370 is enabled. for the same context, I have https://github.com/cilium/cilium/pull/17370#issuecomment-934699604 that there is really some problem between Cilium and Kernel 4.9 - 5.4, I think my PR simply just exposed the problem, appreciate you can  provide feed back

joestringer replied:
My suggestion is to drop this hunk (the pieces in this PR that change `pkg/datapath/linux/requirements.go`) from this PR and move the equivalent code to your other PR. Do you have a counter-proposal?

vincentmli replied:
yes I can drop the hunk in `pkg/datapath/linux/requirements.go`

vincentmli replied:
done

joestringer reviewed (approved):
LGTM.

I notice that this is marked as "draft" and you have not clicked the "ready for review" button. This is likely why folks are not looking at this PR, as we usually use draft status only for under-development PRs that do not need reviewer attention yet.

ti-mo commented on pkg/datapath/linux/probes/probes.go:
```diff
@@ -322,6 +328,16 @@ func (p *ProbeManager) GetHelpers(prog string) map[string]struct{} {
 }
 
+// Probe if kernel support large 1M instructions
+func (p *ProbeManager) ProbeKernelLargeInstruction() error {
```
Why not a simple `(p *ProbeManager) GetMisc() Misc`?

vincentmli replied:
@ti-mo it seems you not just suggesting changing the function name, I used `probeSystemKernelHz()` approach which seems easier for me since I am new to Golang :) for now the `Misc` only has `have_large_insn_limit` item, if I do `(p *ProbeManager) GetMisc() Misc`, it seems there is more code involved, any suggestion to get `have_large_insn_limit` that I am only interested in with our suggestion?

vincentmli replied:
oh, I got it after looking at `GetMapTypes()`
```
+// GetMisc returns information about kernel miscs.
+func (p *ProbeManager) GetMisc() *Misc {
+       return &p.features.Misc
+}
```

the code to query the `Misc`
```
supportedMisc := probes.NewProbeManager().GetMisc()
 if !supportedMisc.HaveLargeInsnLimit {
....
}
```
This PR only include `GetMisc() *Misc `, I think that is what you suggested :)

vincentmli replied:
@ti-mo I pushed the new changes according to your suggestion, could you take a look?

pchaigno reviewed (approved):
LGTM. I agree with Timo's feedback, we should keep the same pattern as we have for helpers, maps and program types.

vincentmli committed:
Add Kernel Misc Probe

Add kernel misc feature probe which include kernel BPF large instruction
support information. It allows Cilium features to probe kernel large instruction
support.

Signed-off-by: Vincent Li <vincent.mc.li@gmail.com>

ti-mo commented on pkg/datapath/linux/probes/probes.go:
```diff
@@ -306,6 +312,11 @@ func (p *ProbeManager) GetMapTypes() *MapTypes {
 
+// GetMisc returns information about kernel misc.
+func (p *ProbeManager) GetMisc() *Misc {
+	return &p.features.Misc
```
Just returning `p.features.Misc` will do. Returning a pointer will unnecessarily allocate on the heap.

vincentmli replied:
ah, good to know that, thanks! done

ti-mo reviewed (approved):
Thanks!
